polyfill/lib/ecmascript.mjs

```diff
@@ -2514,48 +2525,58 @@ export function GetNamedTimeZoneDateTimeParts(id, epochNanoseconds) {
 }
 
 export function GetNamedTimeZoneNextTransition(id, epochNanoseconds) {
-  if (epochNanoseconds.lesser(BEFORE_FIRST_DST)) {
-    return GetNamedTimeZoneNextTransition(id, BEFORE_FIRST_DST);
+  // Optimization: we truncate the instant to the previous millisecond boundary
+  // so that we can do Number math instead of BigInt math. This assumes that
+  // time zone transitions don't happen in the middle of a millisecond.
+  const epochMilliseconds = epochNsToMs(epochNanoseconds, 'trunc');
+  if (epochMilliseconds < BEFORE_FIRST_DST) {
+    return GetNamedTimeZoneNextTransition(id, bigInt(BEFORE_FIRST_DST).multiply(1e6));
   }
+
   // Optimization: the farthest that we'll look for a next transition is 3 years
   // after the later of epochNanoseconds or the current time. If there are no
   // transitions found before then, we'll assume that there will not be any more
   // transitions after that.
-  const now = SystemUTCEpochNanoSeconds();
-  const base = epochNanoseconds.greater(now) ? epochNanoseconds : now;
-  const uppercap = base.plus(bigInt(DAY_NANOS).multiply(366 * 3));
-  let leftNanos = epochNanoseconds;
-  let leftOffsetNs = GetNamedTimeZoneOffsetNanoseconds(id, leftNanos);
-  let rightNanos = leftNanos;
+  const now = DateNow();
+  const base = MathMax(epochMilliseconds, now);
+  const uppercap = base + DAY_MS * 366 * 3;
+  let leftMs = epochMilliseconds;
+  let leftOffsetNs = GetNamedTimeZoneOffsetNanosecondsImpl(id, leftMs);
+  let rightMs = leftMs;
   let rightOffsetNs = leftOffsetNs;
-  while (leftOffsetNs === rightOffsetNs && bigInt(leftNanos).compare(uppercap) === -1) {
-    rightNanos = bigInt(leftNanos).plus(bigInt(DAY_NANOS).multiply(2 * 7));
-    if (rightNanos.greater(NS_MAX)) return null;
-    rightOffsetNs = GetNamedTimeZoneOffsetNanoseconds(id, rightNanos);
+  while (leftOffsetNs === rightOffsetNs && leftMs < uppercap) {
+    rightMs = leftMs + DAY_MS * 2 * 7;
+    if (rightMs > MS_MAX) return null;
+    rightOffsetNs = GetNamedTimeZoneOffsetNanosecondsImpl(id, rightMs);
     if (leftOffsetNs === rightOffsetNs) {
-      leftNanos = rightNanos;
+      leftMs = rightMs;
     }
   }
   if (leftOffsetNs === rightOffsetNs) return null;
   const result = bisect(
-    (epochNs) => GetNamedTimeZoneOffsetNanoseconds(id, epochNs),
-    leftNanos,
-    rightNanos,
+    (epochMs) => GetNamedTimeZoneOffsetNanosecondsImpl(id, epochMs),
+    leftMs,
+    rightMs,
     leftOffsetNs,
     rightOffsetNs
   );
-  return result;
+  return bigInt(result).multiply(1e6);
 }
 
 export function GetNamedTimeZonePreviousTransition(id, epochNanoseconds) {
+  // Optimization: we expand the instant to the next millisecond boundary so
+  // that we can do Number math instead of BigInt math. This assumes that time
+  // zone transitions don't happen in the middle of a millisecond.
+  const epochMilliseconds = epochNsToMs(epochNanoseconds, 'expand');
```

Comment:
The description here is incorrect; we need to _raise_ rather than _expand_ the instant to get correct behavior for instants before the POSIX epoch[^1]—e.g., assuming w.l.o.g. that the last transition of 1969 for some time zone is at epoch nanoseconds T=-1_000_000_000, a request for the previous transition from one nanosecond after that transition [at t₊=-999_999_999] should be treated as equivalent to a request from one millisecond after that transition [at t₊₊=-999_000_000], which would be the result of rounding ns to ms with mode "ceil" (_whereas mode "expand" would produce T and thereby propagate into the incorrect result of the previous transition_).

However, despite the incorrect description, we get the correct results in both this function and GetNamedTimeZoneNextTransition, because `epochNsToMs` does not implement its purported contract—it _claims_ to support "trunc" and "expand", but _actually_ supports "floor" and "ceil":
```js
for (const mode of ["trunc", "expand"]) {
  for (const ns of [123, -123]) {
    console.log(`"${mode}"`, ns, "ns", `(${ns/1e6} ms)`, "→", epochNsToMs(ns, mode), "ms");
  }
}
/* =>
"trunc" 123 ns (0.000123 ms) → 0 ms
"trunc" -123 ns (-0.000123 ms) → -1 ms
"expand" 123 ns (0.000123 ms) → 1 ms
"expand" -123 ns (-0.000123 ms) → 0 ms
*/
```

AFAICT, every use of `epochNsToMs` with mode "trunc" actually wants (and gets) mode "floor", and this sole use with mode "expand" actually wants (and gets) mode "ceil".

[^1]: and in GetNamedTimeZoneNextTransition, we need to _floor_ rather than _truncate_